File: examples/elixir/micro/auth_service.ex
defmodule MyApp.AuthService do
  alias Ecto.Adapters.SQL

  def verify_token(token) do
    query = "SELECT user_id FROM tokens WHERE token = '" <> token <> "'"
    # sink: SQL injection via concatenation
    MyApp.Repo.query(query)
    1
  end

  def run_admin_command(user_id, action) do
    if user_id do
      # sink: command injection via shell concatenation
      System.cmd("sh", ["-c", "notify-admin " <> action])
    end
  end
end

File: examples/lua/micro/auth_service.lua
local M = {}
local luasql = require("luasql.sqlite3")
local db = luasql.sqlite3()

function M.verifyToken(token)
    local query = "SELECT user_id FROM tokens WHERE token = '" .. token .. "'"
    -- sink: SQL injection via concatenation
    db:execute(query)
    return 1
end

function M.runAdminCommand(user_id, cmd)
    if user_id then
        -- sink: command injection via shell concatenation
        os.execute("notify-admin " .. cmd)
    end
end

return M
